### DISCUSSION TITLE
Should SftpSession and Channel be wrapped in an Arc for thread safety or not?

### DISCUSSION THREAD

**Discussion by kanpov** [Eugeny/russh#308] [Q&A] [answered]
Title

**Comment by Eugeny** [comment]
Most russh things are `Send` but not `Sync` if that's what you're asking.

**Reply by kanpov** [reply]
> Most russh things are `Send` but not `Sync` if that's what you're asking.

I'm new to Rust, so not being `Sync` does mean it should be wrapped in an `Arc` to be able to be shared across threads?

**Reply by Eugeny** [reply]
Typically you wouldn't concert yourself with this when writing an async `tokio` based application. Objects such as your russh `Handle`s would never be accessed from multiple threads simultaneously (`Sync`) but instead moved between threads (`Send`).

Thread safety usually only becomes relevant when you're trying to fuse an async library into your thread-based non-async application

**Reply by kanpov** [reply]
> Typically you wouldn't concert yourself with this when writing an async `tokio` based application. Objects such as your russh `Handle`s would never be accessed from multiple threads simultaneously (`Sync`) but instead moved between threads (`Send`).
> 
> Thread safety usually only becomes relevant when you're trying to fuse an async library into your thread-based non-async application

So if I'm writing a library (which is what I'm doing right now) where an `SftpSession` for SFTP and `Channel<Msg>` for SSH are shared between threads, I would need neither `Arc` nor `Mutex`?

**Reply by Eugeny** [reply, marked as answer]
Depends on what you're doing. If you actually share the reference to the structs between your own threads, you need to wrap it in `Arc<Mutex>` like every other non-`Sync` object. If the moving only happens due to async futures migrating between executor threads, you don't need it.
